src/kernels/ipywidgets-message-coordination/cdnWidgetScriptSourceProvider.ts

```diff
@@ -0,0 +1,260 @@
+        if (isWebViewOnline === false) {
+            this.warnIfNoAccessToInternetFromWebView();
+            return {
+                moduleName
+            };
+        }
+        // First see if we already have it downloaded.
+        const key = this.getModuleKey(moduleName, moduleVersion);
+        if (!this.cache.get(key)) {
+            this.cache.set(key, this.getWidgetScriptSourceImplementation(moduleName, moduleVersion));
+        }
+        return this.cache.get(key)!;
+    }
+    protected async generateDownloadUri(
+        moduleName: string,
+        moduleVersion: string,
+        cdn: WidgetCDNs
+    ): Promise<string | undefined> {
+        const cdnBaseUrl = getCDNPrefix(cdn);
+        if (cdnBaseUrl) {
+            return moduleNameToCDNUrl(cdnBaseUrl, moduleName, moduleVersion);
+        }
+        return undefined;
+    }
+
+    protected getModuleKey(moduleName: string, moduleVersion: string) {
+        return `${moduleName}${moduleVersion}`;
+    }
+    protected async getWidgetScriptSourceImplementation(
+        moduleName: string,
+        moduleVersion: string
+    ): Promise<WidgetScriptSource> {
+        traceInfo(
+            `${
+                ConsoleForegroundColors.Green
+            }Searching for Widget Script ${moduleName}#${moduleVersion} using cdns ${this.cdnProviders.join(' ')}`
+        );
+        await this.configureWidgets();
+        if (this.cdnProviders.length === 0) {
+            return { moduleName };
+        }
+        // Try all cdns
+        const uris = await Promise.all(
+            this.cdnProviders.map((cdn) => this.getValidUri(moduleName, moduleVersion, cdn))
+        );
+        const scriptUri = uris.find((u) => u);
+        if (scriptUri) {
+            traceInfo(
+                `${ConsoleForegroundColors.Green}Widget Script ${moduleName}#${moduleVersion} found at URI: ${scriptUri}`
+            );
+            return { moduleName, scriptUri, source: 'cdn' };
+        }
+
+        traceError(`Widget Script ${moduleName}#${moduleVersion} was not found on on any cdn`);
+        this.handleWidgetSourceNotFound(moduleName, moduleVersion).ignoreErrors();
+        return { moduleName };
+    }
+
+    private async getValidUri(moduleName: string, moduleVersion: string, cdn: WidgetCDNs): Promise<string | undefined> {
+        // Make sure CDN has the item before returning it.
+        try {
+            const downloadUrl = await this.generateDownloadUri(moduleName, moduleVersion, cdn);
+            if (downloadUrl && (await this.httpClient.exists(downloadUrl))) {
+                return downloadUrl;
+            }
+        } catch (ex) {
+            traceVerbose(`Failed downloading ${moduleName}:${moduleVersion} from ${cdn}`);
+            return undefined;
+        }
+    }
+    private warnIfNoAccessToInternetFromWebView() {
+        //
+    }
+
+    private async configureWidgets(): Promise<void> {
+        if (this.configuredScriptSources.length !== 0) {
```

Comment:
Moved code related to displaying messages for CDN into the CDN class, instaead of having that elsewhere.